discovery/linode/linode.go

```diff
@@ -0,0 +1,195 @@
+	return NewDiscovery(c, opts.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+// Discovery periodically performs Linode requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client *linodego.Client
+	port   int
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port: conf.Port,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "linode_sd", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	client := linodego.NewClient(
+		&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		},
+	)
+	client.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version))
+	d.client = &client
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"linode",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "Linode",
+	}
+
+	// gather all linode instances
+	instances, err := d.client.ListInstances(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// gather detailed IP address info for all IPs on all linode instances
+	detailedIPs, err := d.client.ListIPAddresses(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range instances {
+		if len(instance.IPv4) == 0 {
+			continue
+		}
+
+		var privateIPv4, publicIPv4, publicIPv6 string
+		for _, detailedIP := range detailedIPs {
+			if detailedIP.LinodeID != instance.ID {
+				continue
+			}
+
+			if detailedIP.Type == "ipv6" {
+				publicIPv6 = detailedIP.Address
+			} else if detailedIP.Type == "ipv4" {
+				if detailedIP.Public {
+					publicIPv4 = detailedIP.Address
+				} else {
+					privateIPv4 = detailedIP.Address
+				}
+			}
+		}
+
+		labels := model.LabelSet{
+			linodeLabelID:          model.LabelValue(fmt.Sprintf("%d", instance.ID)),
+			linodeLabelName:        model.LabelValue(instance.Label),
+			linodeLabelImage:       model.LabelValue(instance.Image),
+			linodeLabelPrivateIPv4: model.LabelValue(privateIPv4),
+			linodeLabelPublicIPv4:  model.LabelValue(publicIPv4),
+			linodeLabelPublicIPv6:  model.LabelValue(publicIPv6),
+			linodeLabelRegion:      model.LabelValue(instance.Region),
+			linodeLabelType:        model.LabelValue(instance.Type),
+			linodeLabelStatus:      model.LabelValue(instance.Status),
+		}
```

Comment:
By looking at the API, we are missing labels about the hypervisor, the group, whether backups are enables. Specs might also make interesting labels.

Reply:
Fine by me, I can add them in 👍